### ISSUE TITLE
Assertion failure false-positives when using OpenGL in audio plugins in OS X

### ISSUE THREAD

**Issue by yairchu** [julianstorer/JUCE#88]
To trigger this bug we need to modify the JUCE demo plugin to use OpenGL with continuous rendering enabled.

Demo plugin modifications:

* Add the juce_opengl module via the projucer
* Add an `OpenGLContext context;` member in `JuceDemoPluginAudioProcessorEditor`.
* Add `context.attachTo (*this); context.setContinuousRepainting (true);` in `JuceDemoPluginAudioProcessorEditor`'s constructor and `context.detach();` in its destructor.

Then debug the plugin in Xcode when running via Traction 7.1.1 (though I believe the behavior is similar across all hosts), and use the plugin as either AU, VST, or VST3. Just open a plugin window and then either close it or switch to a different application. This triggers the assertion error in `checkGLError` about 90% of the time.

Tested on OS X 10.11.6 using Xcode 7.3.1, in 64-bit mode, with either the JUCE master branch or the develop branch.

The following patch which adds some additional checks in `checkGLError` for these cases seems to almost solve the problem but it still occasionally triggers: https://github.com/soundradix/JUCE/commit/7d371d74bfda6312c0af5476c82d438b7c227b17 
When combined with changes making the OpenGL rendering on OS X use `CVDisplayLink` (these new changes are available in soundradix's branch but are not yet thoroughly tested), the above patch appears to solve the issue completely.

**Comment by hogliux** [comment]
I think the suggested checks in checkGLError are really just a workaround for the underlying problem that the openGLContext needs to know when the component is removed from a peer *before* it happens. I'll try to add a workaround....

**Comment by hogliux** [comment]
Can you confirm that this is now fixed on develop?

**Comment by yairchu** [comment]
With the latest develop adding the JUCE demo plugin (unmodified/without OpenGL) in Tracktion as VST or VST3 crashes.

**Comment by hogliux** [comment]
OK Let me know if it works now.

**Comment by yairchu** [comment]
It still crashes as VST and VST3..

Tried to enable address sanitizer and thread sanitizer to see if I get any clues to the cause of the crash.. tsan writes some stuff that seems irrelevant but maybe this line is relevant:

    2016-08-09 12:57:24.735 Tracktion64[14907:5355622] 
    JUCEView_1bdc6b5442e1b823(0x7d2800032820) - NSView not correctly initialized. Did you forget to call super?

?

**Comment by hogliux** [comment]
Can you check if the latest commit on develop fixes the issue?

**Comment by yairchu** [comment]
Yes! Looks like it works now!

**Comment by hogliux** [comment]
Thank you for reporting!

**Comment by yairchu** [comment]
I verified that reverting the fixes for this issue (https://github.com/soundradix/JUCE/commit/193523203642925a44dd565429c1d2b7c7eed5c7) resolves https://github.com/julianstorer/JUCE/issues/120 which was introduced by them.

Applying instead the fix that I previously suggested (https://github.com/soundradix/JUCE/commit/589d7cc2cb913d698dd27628b47e63e6f6eabb3a), on SR's branch which uses CVDisplayLink on macOS, seems to resolve this issue without introducing any other problems afaik so far..
